File: leetcode/3.无重复字符的最长子串/3.无重复字符的最长子串.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	window := map[int32]int{}
	left := 0
	right := 0
	result := 0
	for right < len(s) {
		c := int32(s[right])
		right += 1
		window[c] += 1
		for window[c] > 1 {
			d := int32(s[left])
			left += 1
			window[d] -= 1
		}
		if result < right-left {
			result = right - left
		}
	}
	return result
}

func main() {
	s := "abcabcbb"
	result := lengthOfLongestSubstring(s)
	fmt.Println(result)
}

File: leetcode/3.无重复字符的最长子串/3.无重复字符的最长子串.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = {}
        left, right = 0, 0
        result = 0
        # 扩大窗口，添加字符
        while right < len(s):
            c = s[right]
            right += 1
            window[c] = window.get(c, 0) + 1
            # 字符计数超过1 说明重复，需要缩小窗口
            while window[c] > 1:
                d = s[left]
                left += 1
                window[d] -= 1
            # 当缩小窗口完成时，说明没有重复，符合 最长子串，并和旧长度对比储存
            result = max(result, right - left)
        return result


if __name__ == '__main__':
    s = "abcabcbb"
    result = Solution().lengthOfLongestSubstring(s)
    print(result)
